Fixes part of #9749: Migrate few instances of angular-html-bind

srijanreddy98 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #9749.
2. This PR does the following: Migrates some angular html bind instances that show rte output.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
https://drive.google.com/file/d/1hJooUAgGpvFVTsYBtBIjQddUWaIKEDcE/view?usp=sharing
https://drive.google.com/file/d/1GR1ehPlU0RiD-B9iZcjuSj5q7d7UKw-V/view?usp=sharing
<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

srijanreddy98 committed:
Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

srijanreddy98 committed:
Fix issues

srijanreddy98 committed:
Fix spelling mistakes in comment

Commit:
Merge remote-tracking branch 'upstream/develop' into rte-output

Commit:
Merge with upstream/develop

srijanreddy98 commented:
@vojtechjelinek, @kevintab95, @nithusha21, @DubeySandeep, @EricZLou, @aks681, @Hudda, @iamprayush PTAL!

vojtechjelinek commented on .eslintrc:
```diff
@@ -143,6 +143,7 @@
       "unit-test-utils.ajs.ts",
       "mathjax.directive.ts",
       "math-expression-content-editor.component.ts",
+      "rte-output-display.component.spec.ts",
```
Why is it added here?

DubeySandeep replied:
Instead of disabling catch in the entire file maybe use a bypass flag and state the reason for using the bypass flag through code-comment!

srijanreddy98 replied:
Done!

vojtechjelinek commented on core/templates/components/ck-editor-helpers/ck-editor-4-rte.component.ts:
```diff
@@ -99,6 +119,16 @@ export class CkEditor4RteComponent implements AfterViewInit, OnChanges,
     // components may change without re-rendering each of the components,
     // in such cases, it is sufficient to update the ckeditor instance manually
     // with the latest value.
+    let value = this.value;
```
Maybe just add a comment that will refer to the note at the top of this file. So that folks who just see this part understand it quicker. Ditto below.

srijanreddy98 replied:
Done!

vojtechjelinek reviewed (changes requested):
Thanks! Only two nits.

EricZLou reviewed (approved):
LGTM!

kevintab95 reviewed (approved):
LGTM for codeowners. Thank you @srijanreddy98!

kevintab95 commented:
@srijanreddy98 please assign me once you have attached a testing video, thanks!

iamprayush commented on extensions/objects/templates/math-expression-content-editor.component.ts:
```diff
@@ -80,6 +80,9 @@ export class MathExpressionContentEditorComponent implements OnInit {
       this.value.mathExpressionSvgIsBeingProcessed = true;
     }
     this.valueChanged.emit(this.value);
+    if (this.value && this.value.raw_latex) {
+      this.localValue.label = this.value.raw_latex || '';
+    }
```
We don't need `|| ''` here since we already have an if condition for `this.value.raw_latex` on line 83.

srijanreddy98 replied:
Done!

DubeySandeep commented on core/templates/components/ck-editor-helpers/ck-editor-copy-content.service.ts:
```diff
@@ -122,26 +127,36 @@ export class CkEditorCopyContentService {
       // Look for x-with-value="y" to extract x and y.
-      //  Group 1 (\w+): Any word containing [a-zA-Z0-9_] characters. This is
-      //    the name of the property.
+      //  Group 1 ([\w-]+): Any word containing [a-zA-Z0-9_-] characters. This
+      //    is the name of the property. (Note that this will also catch
+      //    ng-reflect-* which is added by angular).
```
```suggestion
      // Group 1 ([\w-]+): Any word containing [a-zA-Z0-9_-] characters. This
      // is the name of the property. (Note that this will also catch
      // ng-reflect-* which is added by angular).
```

DubeySandeep replied:
Why are we expecting to catch `ng-reflect-*` here? Isn't it only expected to catch interaction-specific attributes?

srijanreddy98 replied:
Yes it is. But then I am also discarding the ng-reflect-* below. I can't precisely recall what went wrong with the old regex but I had to change it. (Line 151-159)

DubeySandeep replied:
On line 144 we are firstly catching these new attributes and then ignoring it on line 158, can we just ignore it through the regex check?

Also, in case it’s a for some “special” case then I’ll would highly suggest writing a code comment here!

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -559,12 +559,14 @@
 /core/templates/services/autoplayed-videos.service*.ts @aks681
 /core/templates/services/external-save.service*.ts @srijanreddy98
 /core/templates/services/external-rte-save.service*.ts @srijanreddy98
+/core/templates/services/oppia-rte-parser.service.ts @srijanreddy98 @aks681
```
```suggestion
/core/templates/services/oppia-rte-parser.service*.ts @srijanreddy98 @aks681
```

Are we expecting this to catch the test files as well?

srijanreddy98 replied:
Yes, we are. Changed it.

DubeySandeep reviewed (changes requested):
@srijanreddy98 Thanks for the PR, I've left  a few minor comments for the codewner files, PTAL!

nithusha21 commented on core/templates/services/oppia-rte-parser.service.spec.ts:
```diff
@@ -0,0 +1,136 @@
+    });
+  });
+
+  // Noninteractive components are easily identified by the oppia-noninteractive
```
What could possibly happen if they use a valid tag but invalid params? Can that cause vulnerabilities?

srijanreddy98 replied:
That part is handled in the rte-output component. Since we are manually typing out all the attributes, random attributes will be ignored. Hence no vuls here.
<img width="1024" alt="Screenshot 2021-11-26 at 10 36 49" src="https://user-images.githubusercontent.com/23002114/143567671-9f1c22d4-f66d-4a44-9f9c-1b9993d36163.png">

Commit:
Merge remote-tracking branch 'upstream/develop' into rte-output

Commit:
Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

srijanreddy98 commented:
@nithusha21 @DubeySandeep @iamprayush @vojtechjelinek PTAL!

srijanreddy98 commented:
@kevintab95 Added videos. PTAL!

vojtechjelinek reviewed (approved):
LGTM!

DubeySandeep commented on core/templates/components/ck-editor-helpers/ck-editor-copy-content.service.ts:
```diff
@@ -73,7 +74,11 @@ export class CkEditorCopyContentService {
+
+      if (
+        parentElement.tagName === this.OUTPUT_VIEW_TAG_NAME ||
+        parentElement.tagName === this.OUTPUT_NG_TAG_NAME
```
Do we still need the ‘ OUTPUT_VIEW_TAG_NAME’ constant?

DubeySandeep reviewed (approved):
@srijanreddy98 I left 2 comments for the 2 code owner  files, PTAL! 

The comments can be resolved in a separate PR!

seanlip commented:
I am merging this because it is needed for the release and @iamprayush's comment is addressed.

@iamprayush Please take a look when you can and confirm that you have no further comments. Thanks!

iamprayush commented:
> I am merging this because it is needed for the release and @iamprayush's comment is addressed.
> 
> @iamprayush Please take a look when you can and confirm that you have no further comments. Thanks!

My comment has been addressed, thanks!
